File: Data Structures/Binary Search Tree/700/700.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func searchBST(root *TreeNode, val int) *TreeNode {
	// If the tree is empty, return nil
	if root == nil {
		return nil
	}

	// If the root's value is equal to the target value, return the root
	if root.Val == val {
		return root
	}

	// If the root's value is greater than the target value, search the left subtree
	if root.Val > val {
		return searchBST(root.Left, val)
	}

	// If the root's value is less than the target value, search the right subtree
	return searchBST(root.Right, val)
}

// Time complexity: O(h), where h is the height of the tree
// Space complexity: O(h), where h is the height of the tree

File: Data Structures/Trees/Binary Search Tree/700/700.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        # If the tree is empty, return None
        if root is None:
            return None

        # If the root's value is equal to the target value, return the root
        if root.val == val:
            return root

        # If the root's value is greater than the target value, search the left subtree
        if root.val > val:
            return self.searchBST(root.left, val)

        # If the root's value is less than the target value, search the right subtree
        return self.searchBST(root.right, val)
    
# time complexity: O(h), where h is the height of the tree
# space complexity: O(h), where h is the height of the tree